File: shop/2019/day4/i.py
s = input()

d = [{} for _ in range(2)]
n = len(s)
for i in range(n):
    d[i % 2][s[i]] = d[i % 2].get(s[i], 0) + 1

odd_cnt = 0
for i in d[0]:
    odd_cnt += d[0][i] % 2

# if odd_cnt ^ (n % 2):
#     print('NO!')
#     exit(0)

odd = ''
if n % 2:
    dd = {}
    for i in d[0]:
        dd[i] = d[0].get(i, 0) - d[1].get(i, 0)

    ones = 0
    gg = ''
    for i in dd:
        if dd[i] != 0 and dd[i] != 1:
            print(dd[i])
            print('NO0')
            exit(0)
        if dd[i] == 1:
            ones += 1
            gg = i
    if ones > 1:
        print('NO1')
        exit(0)
    odd = gg
    d[0][gg] -= 1
if d[0] != d[1]:
    print('NO2')
    exit(0)

ans = ['' for _ in range(n)]
pnt = 0
pnt1 = n - 1
if n % 2:
    ans[n // 2] = odd
    pnt1 += 1

for i in d[0]:
    while d[0][i] > 0:
        ans[pnt] = i
        d[0][i] -= 1
        ans[pnt1] = i
        d[1][i] -= 1
        pnt += 1
        pnt1 -= 1
print(''.join(ans))

File: shop/2019/day4/I.cpp
#include <iostream>
#include <vector>
#include <map>
#include <set>
#include <queue>
#include <algorithm>
#include <string>
#include <cmath>
#include <cstdio>
#include <iomanip>
#include <fstream>
#include <cassert>
#include <cstring>
#include <unordered_set>
#include <unordered_map>
#include <numeric>
#include <ctime>
#include <bitset>
#include <complex>

using namespace std;
int n;
string s;
int cnt[2][26];
inline void init(){
    n = s.size();
}
void odd(){
    for (int i = 0; i < 26; i++){
        if (cnt[0][i] != cnt[1][i]){
            cout << "NO\n";
            return;
        }
    }
    int pnt = 0;
    int pnt1 = n - 1;
    int pnt2 = 0;
    string ans(n, '\0');
    while (pnt < pnt1){
        while (cnt[0][pnt2] == 0)
            pnt2++;
        ans[pnt++] = 'A' + pnt2;
        ans[pnt1--] = 'A' + pnt2;
        cnt[0][pnt2]--;
    }
    cout << ans << '\n';
}
void even(){
    if (n == 3 && s[0] == s[2]){
        cout <<s << '\n';
        return;
    }
    if (n == 3 && s[0] != s[2]){
        cout << "NO\n";
        return;
    }
    int gg = (n / 2) % 2;
    for (int i = 0; i < 26; i++){
        if (cnt[!gg][i] % 2){
            cout << "NO\n";
            return;
        }
    }
    int even = -1;
    for (int i = 0; i < 26; i++){
        if (cnt[gg][i] % 2){
            if (even != -1){
                cout << "NO\n";
                return;
            }
            even = i;
        }
    }
    if (even == -1){
        cout << "NO\n";
    }
    int pnt = 0;
    int pnt1[2];
    pnt1[0] = pnt1[1] = 0;
    string ans(n, '\0');
    cnt[gg][even]--;
    ans[n / 2] = even + 'A';
    while (pnt != n / 2){
        while (cnt[pnt % 2][pnt1[pnt % 2]] == 0)
            pnt1[pnt % 2]++;
        // if (cnt[pnt % 2][pnt])
        ans[n - pnt - 1] = ans[pnt] = pnt1[pnt % 2] + 'A';
        cnt[pnt % 2][pnt1[pnt % 2]] -= 2;
        pnt++;
    }
    cout << ans << '\n';
}
inline void solve(){
    init();
    for (int i = 0; i < n; i++)
        cnt[i % 2][s[i] - 'A']++;
    if (n % 2 == 0)
        odd();
    else
        even();
}


int main(){
	// #ifdef LOCAL
	// 	freopen("I.in", "r", stdin);
	// 	freopen("I.out", "w", stdout);
	// #endif
	
    ios::sync_with_stdio(0);
    cin.tie(0), cout.tie(0);
    cin >> s;
        solve();

}